### ISSUE TITLE
`match` on uninhabited type does not trigger UB in Miri

### ISSUE THREAD

**Issue by theemathas** [rust-lang/rust#142394]
I tried this code:

```rust
#![allow(deref_nullptr)]

enum Never {}

fn main() {
    unsafe {
        match *std::ptr::null::<Result<Never, Never>>() {
            Ok(_) => {
                lol();
            }
            Err(_) => {
                wut();
            }
        }
    }
}

fn lol() {
    println!("lol");
}

fn wut() {
    println!("wut");
}
```

I expected to see this happen: Running this with Miri detects undefined behavior.

Instead, this happened: Running this with Miri outputs "lol", and program exits successfully.

It seems like rustc is discarding the second match arm entirely in the MIR?

<details>
<summary>MIR</summary>

```
// WARNING: This output format is intended for human consumers only
// and is subject to change without notice. Knock yourself out.
// HINT: See also -Z dump-mir for MIR at specific points during compilation.
fn main() -> () {
    let mut _0: ();
    let mut _1: *const std::result::Result<Never, Never>;
    let _2: ();

    bb0: {
        _1 = null::<Result<Never, Never>>() -> [return: bb1, unwind continue];
    }

    bb1: {
        _2 = lol() -> [return: bb2, unwind continue];
    }

    bb2: {
        return;
    }
}

// MIR for lol() and wut() omitted for brevity
```

</details>

### Meta

Reproducible on the playground with `1.89.0-nightly (2025-06-10 1677d46cb128cc8f285d)`

@rustbot labels +A-patterns

**Comment by xizheyin** [comment]
This may because, if either type parameter of `Result` is `Never`, the Rust compiler treats the type as uninhabitable and optimizes away the code, so dereferencing a null pointer doesn't cause a crash; but with `Result<i32, i32>`, the type is inhabited, so dereferencing null will lead to a segmentation fault.

**Comment by theemathas** [comment]
Miri is supposed to run on unoptimized code.

**Comment by saethlin** [comment]
> This may because, if either type parameter of `Result` is `Never`, the Rust compiler treats the type as uninhabitable and optimizes away the code, so dereferencing a null pointer doesn't cause a crash; but with `Result<i32, i32>`, the type is inhabited, so dereferencing null will lead to a segmentation fault.

No. The fact that the pointer here is null is an unnecessary distraction that I didn't realize last night. `Result<Never, Never>` is a ZST, and our documentation clearly states:

> For memory accesses of [size zero](https://doc.rust-lang.org/nightly/nomicon/exotic-sizes.html#zero-sized-types-zsts), every pointer is valid, including the [null](https://doc.rust-lang.org/nightly/std/ptr/fn.null.html) pointer.

Moreover, if you look at the MIR you can see that we never even emit a read. The dereference is fine.

This code has the same strange lowering:
```rust
enum Never {}

fn main(p: *const Result<Never, Never>) {
    unsafe {
        match *p {
            Ok(_) => {
                lol();
            }
            Err(_) => {
                wut();
            }
        }
    }
}

fn lol() {
    println!("lol");
}

fn wut() {
    println!("wut");
}
```

**Comment by kpreid** [comment]
This seems closely related to https://github.com/rust-lang/unsafe-code-guidelines/issues/540, and arguably corresponds to a gap in the definition of what is UB — what are the steps executing a pattern match actually performs? Which one of those steps is is UB when an uninhabited value is *matched* — not ever [_produced_](https://doc.rust-lang.org/reference/behavior-considered-undefined.html#r-undefined.invalid), yet matched?

Here is a simpler program that also passes Miri:

```rust
enum Never {} // also works with ! type

fn main() {
    unsafe {
        match *std::ptr::dangling::<Never>() {
            _ => println!("never happened"),
        }
    }
}
```

This one lacks the puzzle of “which of the match arms should be run? neither, of course”, though, and one could perhaps argue that it *shouldn’t* be UB since there is no branch and Rust prefers not to have entirely gratuitous UB. Yet, what semantics justifies that?

**Comment by kpreid** [comment]
Here are a couple of possible high-level statements of where the UB should be (these do not reflect the current documented semantics of Rust, but could in the future):

1. The *discriminant* of `Result<Never, Never>` is, itself, uninhabited. Therefore, reading it to match it is “producing an invalid value” UB for [`std::mem::Discriminant<Result<Never, Never>>`](https://doc.rust-lang.org/std/mem/struct.Discriminant.html) or an equivalent internal construct.
2. A match arm whose pattern is uninhabited should be unreachable; its body should never execute. Therefore, none of the arms of these `match`es match, and control flow falls off the end of the `match`, which should be defined as UB.

**Comment by RalfJung** [comment]
I agree this should be UB -- I'd have expected a discriminant read that causes UB here. But somehow we don't get one?
Cc @Nadrieril

**Comment by RalfJung** [comment]
The MIR is already odd immediately after being built:
```
fn main(_1: *const Result<Never, Never>) -> () {
    debug p => _1;
    let mut _0: ();
    let _2: ();
    let _3: ();

    bb0: {
        PlaceMention((*_1));
        falseEdge -> [real: bb3, imaginary: bb1];
    }

    bb1: {
        StorageLive(_3);
        _3 = wut() -> [return: bb5, unwind: bb7];
    }

    bb2: {
        FakeRead(ForMatchedPlace(None), (*_1));
        unreachable;
    }

    bb3: {
        StorageLive(_2);
        _2 = lol() -> [return: bb4, unwind: bb7];
    }
```
Note the entire lack of a discriminant check.

**Comment by Nadrieril** [comment]
Yep, this is something we're about to fix. Today we skip discriminant reads for enums based on the presence of empty variants; this was considered incorrect and t-lang FCP-approved a fix [here](https://github.com/rust-lang/rust/pull/138961) (which is apparently waiting on me?? oops).

**Comment by RalfJung** [comment]
I recall having discussions around this, but... that PR mentions closures and there are no closures involved here?

**Comment by Nadrieril** [comment]
Ah yeah you're right. That PR is only the first step towards fixing that, but we intend to fix that eventually.

**Comment by zachs18** [comment]
Similar to kpreid's example with a single `_` arm, this example passes Miri and both asserts ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=d6adf688fc85ea105dd8c8734a5e7e5a)):
```rust
pub enum Empty {}
#[repr(C)]
pub struct Foo(u32, u32, u32, Empty);
pub unsafe fn foo(x: *const Foo) -> u32 {
    unsafe { match *x {
        Foo(0, val, _, _) => val,
        Foo(_, _, val, _) => val,
    } }
}
fn main() {
    unsafe {
        let x: [u32; 3] = [0, 5, 6];
        assert_eq!(foo((&raw const x).cast()), 5);
        let x: [u32; 3] = [1, 5, 6];
        assert_eq!(foo((&raw const x).cast()), 6);
    }
}
```

Should this example be UB?

(Should "matching on uninhabited `struct`s" be split out from this issue (or maybe the cases without a discriminant read of an uninhabited enum variant in general, so including kpreid's `_` example?), and this issue focused on `enum`s (or cases with discriminant reads of uninhabited variants)?)

<details> <summary> A similar <code>#[repr(u32)]</code> enum example </summary>

This example, seemingly very similar to the above `struct` example, passes Miri but fails the second asssert (presumably for the same reasons as the OP example) ([playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=483c386da4fbabe88f24b53bf895ea9d)):

```rust
pub enum Empty {}
#[repr(u32)]
pub enum Bar {
    A(u32, u32, Empty) = 0,
    B(u32, u32, Empty) = 1,
}
pub unsafe fn bar(x: *const Bar) -> u32 {
    unsafe { match *x {
        Bar::A(val, _, _) => val,
        Bar::B(_, val, _) => val,
    } }
}
fn main() {
    unsafe {
        let x: [u32; 3] = [0, 5, 6];
        assert_eq!(bar((&raw const x).cast()), 5);
        let x: [u32; 3] = [1, 5, 6];
        assert_eq!(bar((&raw const x).cast()), 6);
    }
}
```


The MIR for `fn bar` essentially contains only  `_0 = copy (((*_1) as A).0: u32);` after some null and alignment checks, so as above there is no discriminant check.

Should reading the tag of an uninhabited variant in a primitive-representation enum be UB, even if we don't read an uninhabited field? It seems like it should, given that Miri reports UB on the first match if you replace `B`'s `Empty` field with `()`, but I wonder if it would ever be useful for it not to be UB for specifically primitive-representation enums, specifically when matching on an "unsafe place" (from a raw pointer, not by-value or from a reference)

</details>

**Comment by Nadrieril** [comment]
@pitust fair enough, I guess it's less obviously wrong than the OP. Hopefully this gets settled in the Reference soon.

**Comment by RalfJung** [comment]
There's been plenty of discussion of that question, in particular in the context of https://github.com/rust-lang/rust/issues/120141, but no conclusion yet. 

But I think there is fairly broad agreement that *even if* you are allowed to write to that `u32` field, you are definitely *not* allowed to write the `Ok` discriminant. Writing that discriminant requires first initializing all fields, which is impossible for this case. That means it is impossible to ever read that discriminant, so you can't ever successfully match against `Ok`.
